strace_process_tree.py

```diff
@@ -27,18 +44,80 @@
 __licence__ = 'GPL v2 or v3'  # or ask me for MIT
 
 
-Tree = namedtuple('Tree', 'trunk, fork, end, space')
+class Tree(NamedTuple):
+    trunk: str
+    fork: str
+    end: str
+    space: str
 
 
-class Theme(object):
+Style = Callable[[Any], str]
+Mogrifier = Callable[[str], str]
+Pid = int
+TimeSeconds = float
+SemanticStyle = Literal["tree_style", "pid", "process", "time_range"]
+VisualStyle = Literal["normal", "red", "green", "blue"]
 
-    default_styles = dict(
-        tree_style='normal',
-        pid='red',
-        process='green',
-        time_range='blue',
-    )
 
+def NoopStyle() -> Style:
+    # mypy complains that this is used with a non-method but we are creating a method with this.
+    @staticmethod  # type: ignore
+    def _no_format(text: Any) -> str:
+        return text or ''
+
+    return _no_format
+
+
+ANSI_RESET_CODE = '\033[m'
+
+
+def AnsiStyle(enter_code: str) -> Style:
+    # mypy complains that this is used with a non-method but we are creating a method with this.
+    @staticmethod  # type: ignore
+    def _format(text: Any) -> str:
+        if not text:
+            return ''
+        return f'{enter_code}{text}{ANSI_RESET_CODE}'
+
+    return _format
+
+
+class Theme(ABC):
+    tree: Tree
+
+    normal = NoopStyle()
+    red: Style
+    green: Style
+    blue: Style
+
+    tree_style: Style
+    pid: Style
+    process: Style
+    time_range: Style
+
+    def __init__(self, tree: Tree, styles: Dict[SemanticStyle, VisualStyle] = {}) -> None:
+        self.tree = tree
+
+        # Set up semantic styles
+        self.tree_style = getattr(self, styles.get("tree_style", "normal"))
+        self.pid = getattr(self, styles.get("pid", "red"))
+        self.process = getattr(self, styles.get("process", "green"))
+        self.time_range = getattr(self, styles.get("time_range", "blue"))
```

Comment:
Clearly I went overboard with the Theme magic.  It was an experiment to see how to add configurable color highlighting without affecting too much of the code.  I still like passing around Theme objects and invoking `theme.thing()` before printing things.  I regret abusing `Theme.__new__` instead of writing a standalone `def pick_theme(unicode: Optional[bool], color: Optional[bool]) -> Theme`.

I'm undecided about the magic `__getattr__`.  I'd like to study the problem of mypy dealing with `__getattr__` in greater detail, which will require me to find a block of time.

Reply:
The first issue is that `self.ctlseq` is not defined in `Theme` so `Theme.__getattr__()` cannot refer to it.

I can move `Theme.__getattr__()` to `AnsiTheme.__getattr__()` but then it will not be possible to call it on `Theme` values.

I can resolve that by having `Theme.__new__()` return `Union[AnsiTheme, PlainTheme` instead of `Theme` which will probably satisfy `mypy` but the dynamic declaration of style properties still makes the code harder to reason about than necessary. Which is why I propose moving the style properties to class/object instantiation.

Reply:
The first issue is that `self.ctlseq` is not defined in `Theme` so `Theme.__getattr__()` cannot refer to it.

I can move `Theme.__getattr__()` to `AnsiTheme.__getattr__()` but then it will not be possible to call it on `Theme` values.

I can resolve that by having `Theme.__new__()` return `Union[AnsiTheme, PlainTheme` instead of `Theme` which will probably satisfy `mypy` but the dynamic declaration of style properties still makes the code harder to reason about than necessary. Which is why I propose moving the style properties to class/object instantiation.

Reply:
The first issue is that `self.ctlseq` is not defined in `Theme` so `Theme.__getattr__()` cannot refer to it.

I can move `Theme.__getattr__()` to `AnsiTheme.__getattr__()` but then it will not be possible to call it on `Theme` values.

I can resolve that by having `Theme.__new__()` return `Union[AnsiTheme, PlainTheme` instead of `Theme` which will probably satisfy `mypy` but the dynamic declaration of style properties still makes the code harder to reason about than necessary. Which is why I propose moving the style properties to class/object instantiation.

Reply:
The first issue is that `self.ctlseq` is not defined in `Theme` so `Theme.__getattr__()` cannot refer to it.

I can move `Theme.__getattr__()` to `AnsiTheme.__getattr__()` but then it will not be possible to call it on `Theme` values.

I can resolve that by having `Theme.__new__()` return `Union[AnsiTheme, PlainTheme` instead of `Theme` which will probably satisfy `mypy` but the dynamic declaration of style properties still makes the code harder to reason about than necessary. Which is why I propose moving the style properties to class/object instantiation.

Reply:
Would something like

```python
StyleName = str
ColorName = str

class StyleProperty:
    name: StyleName
    def __get__(...): ...
    def __set_name__(...): ...

class Theme:
    pid = StyleProperty()
    process = StyleProperty()
    ....
    reset: str = ''
    colors: Dict[ColorName, str] = {}
    styles: Dict[StyleName, ColorName] = {}

class PlainTheme(Theme):
    pass

class AnsiTheme(Theme):
    reset = ''
    colors = {
        'red': ...
    }
    styles = {
        'pid': 'red',
    }

def pick_theme(...) -> Theme:
    ...
```

work?  With some magic in StyleProperty.__get__ that needs to return a callable (sort of like a custom bound method) that remembers which theme it came from, and the name of the style property, and can look up `theme.colors.get(theme.styles.get(style_name), '')` and then wrap the argument between it and `theme.reset`?

Reply:
I do not see a benefit over the current proposal.

Other than `pick_theme()` being a function rather than class method of `ThemeFactory`. But then we would need to either define the helpers like `should_use_color` inside the `pick_method` function, or move them outside, losing the grouping and encapsulation provided by the class.